### PR TITLE
Engine: Add the `wait` argument to `submit`

### PR DESCRIPTION
**Pull request by sphuber**
For demos and tutorials, often in interactive notebooks, it is often preferred to use `run` instead of `submit` because in this way the cell will block until the process is done. The cell blocking will signal to the user that the process is still running and as soon as it returns it is immediately clear that the results are ready. With `submit` the cell returns immediately, but the user will now have to resort to manually checking when the process is done. Solutions are to instruct the user to call `verdi process list` manually (which they will have to do repeatedly) or implement some automated loop that checks for the process to terminate.

However, using `run` has downsides as well, most notably that the process will be lost if the notebook gets disconnected. For processes that are expected to run longer, this can be really problematic, and so `submit` will have to be used regardless.

Here, the `wait` argument is added to `submit`. Set to `False` by default to keep current behavior, when set to `True`, the function will mimic the behavior of `run` and only return when the process has terminated at which point the node is returned. A `REPORT` log message is emitted each time the state of the process is checked in intervals of `wait_interval`.

### CODE REVIEW

File: aiida/engine/launch.py [hunk 1]
```diff
@@ -118,8 +119,38 @@ def submit(process: TYPE_SUBMIT_PROCESS, **inputs: Any) -> ProcessNode:
...
+        return node
+
+    while not node.is_terminated:
+        time.sleep(wait_interval)
+        LOGGER.report(f'Process<{node.pk}> has not yet terminated, current state is `{node.process_state}`.')
+
+    return node
+
+
+def await_processes(nodes: t.Sequence[ProcessNode], sleep_interval: int = 1) -> None:
```
Comment (mbercx) [inline aiida/engine/launch.py:134]: Maybe also call it `wait_interval` here, to be consistent with the `submit` argument?

File: docs/source/topics/processes/usage.rst [hunk 2]
```diff
@@ -350,6 +350,28 @@ The function will submit the calculation to the daemon and immediately return co
 .. warning::
     For a process to be submittable, the class or function needs to be importable in the daemon environment by a) giving it an :ref:`associated entry point<how-to:plugin-codes:entry-points>` or b) :ref:`including its module path<how-to:faq:process-not-importable-daemon>` in the ``PYTHONPATH`` that the daemon workers will have.
 
+.. tip::
+    Use ``wait=True`` when calling ``submit`` to wait for the process to complete before returning the node.
+    This can be useful for tutorials and demos in interactive notebooks where the user should not continue before the process is done.
+    One could of course also use ``run`` (see below), but then the process would be lost if the interpreter gets accidentally shut down.
+    By using ``submit``, the process is run by the daemon which takes care of saving checkpoints so it can always be restarted in case of problems.
+    If you need to launch multiple processes in parallel and want to wait for all of them to be finished, simply use ``submit`` with the default ``wait=True`` and collect the returned nodes in a list.
```
Comment (mbercx) [inline docs/source/topics/processes/usage.rst:358]:
```suggestion
    If you need to launch multiple processes in parallel and want to wait for all of them to be finished, simply use ``submit`` with the default ``wait=False`` and collect the returned nodes in a list.
```

File: docs/source/topics/processes/usage.rst [hunk 3]
```diff
@@ -350,6 +350,28 @@ The function will submit the calculation to the daemon and immediately return co
...
+    You can then pass them to :func:`aiida.engine.launch.await_processes` which will return once all processes have terminated::
+
+        .. code:: python
```
Comment (mbercx) [inline docs/source/topics/processes/usage.rst:361]: You've added extra indentation here, which means this line will already be considered a code block.

Screenshot 2023-10-26 at 22 37 41

File: docs/source/topics/processes/usage.rst [hunk 4]
```diff
@@ -350,6 +350,28 @@ The function will submit the calculation to the daemon and immediately return co
...
+
+            nodes = []
+
+            for i in range(5):
+                node = submit(...)
+                nodes.append(node)
+
+            await_processes(nodes, sleep_interval=10)
+
+    The ``await_processes`` function will loop every ``sleep_interval`` seconds and check whether all processes (represented by the ``ProcessNode``s in the ``nodes`` list) have terminated.
```
Comment (mbercx) [inline docs/source/topics/processes/usage.rst:372]: I think the extra `s` after the double backticks is causing trouble here.

Screenshot 2023-10-26 at 22 39 59

**Review by mbercx — CHANGES_REQUESTED** [review]
Thanks @sphuber! Looks good, just a few small comments and it's ready to go.

File: docs/source/topics/processes/usage.rst (hunk 4 above)
Comment (sphuber) [reply docs/source/topics/processes/usage.rst:372]: ah yeah I forgot. Restructured text cannot parse this properly. Absolute garbage...

**Review by unkcpz — COMMENTED** [review]
Looks all perfect to me. 
I'll let @mbercx approve it.

File: docs/source/topics/processes/usage.rst [hunk 5]
```diff
@@ -350,6 +350,29 @@ The function will submit the calculation to the daemon and immediately return co
...
 
+.. tip::
+    Use ``wait=True`` when calling ``submit`` to wait for the process to complete before returning the node.
+    This can be useful for tutorials and demos in interactive notebooks where the user should not continue before the process is done.
+    One could of course also use ``run`` (see below), but then the process would be lost if the interpreter gets accidentally shut down.
+    By using ``submit``, the process is run by the daemon which takes care of saving checkpoints so it can always be restarted in case of problems.
+    If you need to launch multiple processes in parallel and want to wait for all of them to be finished, simply use ``submit`` with the default ``wait=False`` and collect the returned nodes in a list.
+    You can then pass them to :func:`aiida.engine.launch.await_processes` which will return once all processes have terminated::
```
Comment (mbercx) [inline docs/source/topics/processes/usage.rst:359]:
```suggestion
    You can then pass them to :func:`aiida.engine.launch.await_processes` which will return once all processes have terminated:
```

I think the extra colon is still causing trouble here.

Screenshot 2023-10-27 at 17 35 40

**Review by mbercx — DISMISSED** [review]
Just one final fix needed for the Restructured text, it seems. ^^ I'll already approve, thanks again @sphuber!

File: docs/source/topics/processes/usage.rst (hunk 5 above)
Comment (sphuber) [reply docs/source/topics/processes/usage.rst]: Thanks for catching that. I tested the docs locally and it passed, but didn't bother to look at the actual output. Funnily enough, with just a single colon VSCode actually has incorrect syntax highlighting and lights up the whole phrase starting from `:func:` onwards. This was fixed by the double colon, but clearly that is not what restructured text wants.

I also noted that I forgot to update `sleep_interval` to `wait_interval` which I corrected now

**Pull request merged**